feat(synchronization): Implement a global advisory locking mechanism using Redis.

karpandsmeargle opened the pull request:
Implements a very rudimentary global or distributed lock on redis; calling lock() on a thread will block the thread until it can acquire the lock, at which point it is the only thread in **any shard** holding the lock.

Simultaneously to the critical section, a separate thread will run, refreshing the expiry on the entry provided this shard still owns it. On the end of the critical section (unlock), the lock entry will be deleted and this separate thread will be stopped; it could live beyond the critical section, but this at most causes one more refresh (which has no particular negative impact) provided that the same shard acquires the lock.

Keyspace notifications have been enabled to wake threads waiting on the redis database entry to acquire the lock.

Tested on the [main plugin repo](https://github.com/TeamMonumenta/monumenta-plugins/blob/d3caf1268b1861c1d34bbb9064201fcc524862ca/plugins/paper/src/main/java/com/playmonumenta/plugins/commands/TestRedisLockCommand.java), on dev2, dev3, dev4.

karpandsmeargle committed:
Implement RedisLock

karpandsmeargle committed:
Make RedisLockException extend RuntimeException

karpandsmeargle committed:
(Temporary?) enable keyspace notifications

karpandsmeargle committed:
Use Redis WATCH to make sure shared resource does not change

Njol commented on plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java:
```diff
@@ -0,0 +1,245 @@
+	private Lock mInternalAccessLock;
+
+	private Condition mIsIntraLockedCondition;
+	private boolean mIsIntraLocked;
```
This field should be `volatile`, as well as any other modifiable field. Basically, you should only have either `volatile` or  `final` fields in concurrent contexts in Java.

karpandsmeargle committed:
Fix volatile and remove a boolean

NickNackGus commented on plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java:
```diff
@@ -0,0 +1,296 @@
+		return Optional.ofNullable(
+			RedisAPI.getInstance()
+				.sync()
+				.set(mKeyName, ConfigAPI.getShardName(), SetArgs.Builder.nx().px(mTimeoutMS))
```
The shard name is not a sufficiently unique lock value for our uses. Multiple players may try to claim a lock on a market entry from the same shard, for instance. In the mailbox branch, I use a UUID, which will never conflict within our lifetimes.

karpandsmeargle replied:
Since only one thread on the shard can attempt to claim the lock at a time due to the intra-lock, the shard name is sufficient.

NickNackGus commented on plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java:
```diff
@@ -0,0 +1,296 @@
+								.multi();
+							RedisAPI.getInstance()
+								.async()
+								.pexpire(mKeyName, mTimeoutMS);
```
Is there any reason to use a multi/exec block for a single command? This seems like something that could be done directly instead.

karpandsmeargle replied:
This is to leverage the WATCH (performed in `watchThenGetLockingShard`; also see the comment above this) so that the expire doesn't happen if the shard changes.

NickNackGus reviewed (changes requested):
Note that Bukkit has been complaining about us not cancelling threads at shutdown - this needs some additional code for `onDisable` in the plugin code.

In addition to my other comments, this could use some examples in the example plugin for this repo. Some things to demonstrate:
- Example code for how to properly handle a "long" task, possibly using a one second `Thread.sleep()` for demo purposes on an async thread, and how to handle lock exceptions and releasing the lock in both normal and error conditions. It may be best to tie this to a command so that you can test/demo using the same lock multiple times from the same shard.
- Blocking attempt to acquire a lock with a long/infinite timeout
- Non-blocking attempt to acquire a lock (or otherwise comment about using a "reasonably short" timeout, which should be long enough to handle requests for a lock across geographic regions as is required on play/volt)

karpandsmeargle committed:
Make executors static

karpandsmeargle committed:
Improve concurrency in unlock method; handle failed refresh task scheduling by releasing locks before rethrow

karpandsmeargle committed:
Convert to name-based intra lock

karpandsmeargle committed:
Rename to ReentrantLock

karpandsmeargle committed:
Add Reentrant implementation

karpandsmeargle committed:
doc(redislock): Provide full documentation of all public-facing components.
fix(redislock): Re-retrieve MutableLockData after condition variable await drops lock

karpandsmeargle committed:
Rename lockOwner to lockingThread; fix some optional smells

karpandsmeargle committed:
Whitespace fixes

karpandsmeargle committed:
Write example

karpandsmeargle committed:
fix(build): Add deploy tasks, generate plugin.yml, generate bnugee.yml

karpandsmeargle committed:
fix(redislock): Provide more granular timestamps in toy example to allow correct sorting

karpandsmeargle committed:
Reorder executor shutdown

karpandsmeargle committed:
feat(api): Add connection pooling to the API

karpandsmeargle committed:
fix(lock): Prevent transactions from affecting each other using connection pool

karpandsmeargle committed:
test(lock): Add tests to ensure exceptions are properly thrown on incorrect unlocks.

karpandsmeargle committed:
style(lock): Fix style in ExceptionsExampleCommand

karpandsmeargle committed:
rename(lock): ExceptionsExampleCommand -> ExceptionsToyExampleCommand

karpandsmeargle committed:
test(lock): Add more examples

karpandsmeargle committed:
style(lock): Rename timeoutMS to redisExpiryMS

karpandsmeargle committed:
Merge branch 'master' into master

karpandsmeargle committed:
Merge branch 'master' of github.com:TeamMonumenta/monumenta-redis-sync

karpandsmeargle committed:
fix(javadoc): Change bold tag to b
